### PR TITLE
[SAP] Added sap_affinity_filter.py

### PR DESCRIPTION
**Pull request by hemna**
This patch adds the SAPDifferentBackendFilter and SAPSameBackendFilter
classes.  These classes are the affinity/anti-affinity classes to check
and ensure that the same/different netapp backends are used during
volume provisioning requests.

The same cinder api is used for passing in the scheduler hint.
cinder volume create --name foo --hint same_host=<UUID>
cinder volume create --name foo --hint different_host=<UUID>

If a datastore has a custom attribute set on it with the name of
netapp_fqdn then that value is compared in the filters to ensure
the same/different datastore is chosen, depending on the scheduler
hint passed in.

### CODE REVIEW

File: cinder/scheduler/filters/sap_affinity_filter.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        # or single uuid.  The checks here is to make sure every single string
+        # in the list looks like a uuid, otherwise, this filter will fail to
+        # pass.  Note that the filter does *NOT* ignore string doesn't look
+        # like a uuid, it is better to fail the request than serving it wrong.
+        if isinstance(affinity_uuids, list):
+            for uuid in affinity_uuids:
+                if uuidutils.is_uuid_like(uuid):
+                    continue
+                else:
+                    return False
```
Comment (fwiesel) [inline cinder/scheduler/filters/sap_affinity_filter.py:57]: I would say
```python
if uuidutils.is_uuid_like(uuid):
    return False
```

File: cinder/scheduler/filters/sap_affinity_filter.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        # reset the lookup for this request.  It can change between requests.
+        self.backend_fqdn_lookup = {}
+
+        # First we need to get all of the pools and populate a lookup for
+        # the datastore backend fqdn custom attributes, if any.
+        scheduler_hints = filter_properties.get('scheduler_hints') or {}
+        affinity_uuids = scheduler_hints.get(self.hint_key, [])
+        # filter_obj_list is a generator, so we need to copy
+        # each entry for looping again.
+        obj_list = []
```
Comment (fwiesel) [inline cinder/scheduler/filters/sap_affinity_filter.py:88]: Personally, I prefer the shorter branch first.
```python
if not affinity_uuids:
    obj_list = filter_obj_list
else:
    obj_list = []
    ....
```

File: cinder/scheduler/filters/sap_affinity_filter.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        if affinity_uuids:
+            # Build the backend_fqdn_lookup, since we will need it this request
+            objs = list(filter_obj_list)
+            for obj in objs:
```
Comment (fwiesel) [inline cinder/scheduler/filters/sap_affinity_filter.py:92]: Why are you not simply iterating over `filter_obj_list`? In line 102, you set `obj_list = filter_obj_list` and iterate in 104 over it, so it should be possible, no?

**Review by fwiesel — APPROVED** [review]
Looks good to me, except for some minor stylistic things I would have done differently, but no blocker from my side.

File: cinder/scheduler/filters/sap_affinity_filter.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
+# Copyright 2014, eBay Inc.
```
Comment (fwiesel) [inline cinder/scheduler/filters/sap_affinity_filter.py:1]: Ah, probably that should be "2022, SAP SE"

Comment (hemna) [reply cinder/scheduler/filters/sap_affinity_filter.py]: fixed

**Pull request merged**